File: mac-app/Tests/QuartzTeachersTests/FolderPathRewriterTests.swift
import XCTest
@testable import QuartzTeachers

/// Pointing qualified links at a folder's new name — and, just as important,
/// leaving alone the links that need nothing.
@MainActor
final class FolderPathRewriterTests: XCTestCase {

    // MARK: - What must change

    func testAQualifiedWikiLinkFollowsTheFolder() {
        XCTAssertEqual(
            FolderPathRewriter.rewriting("See [[Tasks/Quiz 1]] today.", folderNamed: "Tasks", to: "Assessments"),
            "See [[Assessments/Quiz 1]] today."
        )
    }

    func testATransclusionFollowsTheFolder() {
        XCTAssertEqual(
            FolderPathRewriter.rewriting("![[Tasks/diagram.png]]", folderNamed: "Tasks", to: "Assessments"),
            "![[Assessments/diagram.png]]"
        )
    }

    /// The alias and the heading are the teacher's own words, and a rewriter
    /// that touched them would quietly edit their prose.
    func testAnAliasAndAHeadingAreLeftAlone() {
        XCTAssertEqual(
            FolderPathRewriter.rewriting(
                "[[Tasks/Quiz 1#Marking|the Tasks quiz]]", folderNamed: "Tasks", to: "Assessments"
            ),
            "[[Assessments/Quiz 1#Marking|the Tasks quiz]]"
        )
    }

    /// Obsidian writes a full vault path when a name is ambiguous, so the
    /// folder is not always the first segment.
    func testAFolderDeepInAPathIsFound() {
        XCTAssertEqual(
            FolderPathRewriter.rewriting(
                "[[ICS3U/section1/Tasks/Quiz 1]]", folderNamed: "Tasks", to: "Assessments"
            ),
            "[[ICS3U/section1/Assessments/Quiz 1]]"
        )
    }

    func testAMarkdownLinkFollowsTheFolder() {
        XCTAssertEqual(
            FolderPathRewriter.rewriting(
                "[the quiz](Tasks/Quiz 1.md)", folderNamed: "Tasks", to: "Assessments"
            ),
            "[the quiz](Assessments/Quiz 1.md)"
        )
    }

    /// A percent-encoded segment must come back percent-encoded, or the link
    /// stops resolving — which would be a rename that broke the very links it
    /// set out to keep working.
    func testAPercentEncodedSegmentStaysEncoded() {
        XCTAssertEqual(
            FolderPathRewriter.rewriting(
                "[the quiz](Extra%20Tasks/Quiz%201.md)", folderNamed: "Extra Tasks", to: "Extra Assessments"
            ),
            "[the quiz](Extra%20Assessments/Quiz%201.md)"
        )
    }

    /// Obsidian resolves names without regard to case, so a link written
    /// `tasks/` points at the `Tasks` folder and has to follow it.
    func testMatchingIgnoresCase() {
        XCTAssertEqual(
            FolderPathRewriter.rewriting("[[tasks/Quiz 1]]", folderNamed: "Tasks", to: "Assessments"),
            "[[Assessments/Quiz 1]]"
        )
    }

    // MARK: - What must NOT change

    /// The common case, and the reason a folder rename is far less dangerous
    /// than a page rename: Obsidian resolves a bare page name by searching the
    /// vault, so moving the folder leaves the link working.
    func testABarePageLinkIsUntouched() {
        let text: String = "See [[Quiz 1]] today."
        XCTAssertEqual(FolderPathRewriter.rewriting(text, folderNamed: "Tasks", to: "Assessments"), text)
    }

    /// The last segment is the page, so a page that happens to be CALLED
    /// `Tasks` survives a rename of the folder.
    func testAPageNamedAfterTheFolderIsUntouched() {
        let text: String = "[[Handbook/Tasks]]"
        XCTAssertEqual(FolderPathRewriter.rewriting(text, folderNamed: "Tasks", to: "Assessments"), text)
    }

    /// A segment must be the WHOLE name. Rewriting on a substring would rename
    /// a folder the teacher never touched.
    func testAFolderWhoseNameMerelyContainsTheOldOneIsUntouched() {
        let text: String = "[[Extra Tasks/Quiz 1]] and [[Tasks Archive/Old quiz]]"
        XCTAssertEqual(FolderPathRewriter.rewriting(text, folderNamed: "Tasks", to: "Assessments"), text)
    }

    func testPlainProseMentioningTheFolderIsUntouched() {
        let text: String = "Everything in Tasks/ counts for marks."
        XCTAssertEqual(FolderPathRewriter.rewriting(text, folderNamed: "Tasks", to: "Assessments"), text)
    }

    /// Found by adversarial review, not by use, and it was a real bug: the
    /// segment walk is blind to what a path MEANS, so a folder called `Tasks`,
    /// `Resources` or `Notes` used to repoint every external link whose URL
    /// happened to carry that segment.
    func testAWebAddressIsNeverRewritten() {
        let text: String = "See [the handout](https://example.com/Tasks/handout.pdf) and "
                         + "[more](http://school.example/Tasks/x)."
        XCTAssertEqual(FolderPathRewriter.rewriting(text, folderNamed: "Tasks", to: "Assessments"), text)
        XCTAssertEqual(FolderPathRewriter.countReferences(to: "Tasks", in: text), 0)
    }

    func testAnAbsolutePathOnThisMachineIsNeverRewritten() {
        let text: String = "[the file](/Users/teacher/Tasks/notes.md)"
        XCTAssertEqual(FolderPathRewriter.rewriting(text, folderNamed: "Tasks", to: "Assessments"), text)
    }

    /// Tested against the SHAPE of a scheme rather than a list of schemes: the
    /// list is open, and a missed one silently rewrites somebody's link.
    func testOtherSchemesAreLeftAloneToo() {
        let text: String = "[vault](obsidian://open?file=Tasks/x) [f](file:///Tasks/y.md)"
        XCTAssertEqual(FolderPathRewriter.rewriting(text, folderNamed: "Tasks", to: "Assessments"), text)
    }

    /// The guard must not overreach: a relative path is still rewritten, and a
    /// course whose own folders contain a colon is impossible (the rename sheet
    /// refuses one).
    func testARelativePathIsStillRewritten() {
        XCTAssertEqual(
            FolderPathRewriter.rewriting("[q](./Tasks/Quiz 1.md)", folderNamed: "Tasks", to: "Assessments"),
            "[q](./Assessments/Quiz 1.md)"
        )
    }

    func testRenamingToTheSameNameChangesNothing() {
        let text: String = "[[Tasks/Quiz 1]]"
        XCTAssertEqual(FolderPathRewriter.rewriting(text, folderNamed: "Tasks", to: "Tasks"), text)
    }

    // MARK: - Counting

    func testCountingFindsOnlyQualifiedLinks() {
        let text: String = "[[Tasks/Quiz 1]] [[Quiz 2]] [the third](Tasks/Quiz 3.md) [[Handbook/Tasks]]"
        XCTAssertEqual(FolderPathRewriter.countReferences(to: "Tasks", in: text), 2)
    }

    func testCountingIsZeroWhenNothingPointsIn() {
        XCTAssertEqual(
            FolderPathRewriter.countReferences(to: "Tasks", in: "[[Quiz 1]] and [[Concepts/Loops]]"), 0
        )
    }
}

File: windows-app/Plantoir.Tests/FolderPathRewriterTests.cs
using Plantoir.Core.Models;
using Xunit;

namespace Plantoir.Tests;

/// <summary>
/// Pointing qualified links at a folder's new name.
///
/// <para>The "must NOT change" half is the important one: a rewriter that
/// matched substrings would rename folders the teacher never touched, and one
/// that was blind to web addresses would repoint a link at somebody else's
/// website.</para>
/// </summary>
public class FolderPathRewriterTests
{
    private static string Rewrite(string text) => FolderPathRewriter.Rewritten(text, "Tasks", "Assignments");

    // ------------------------------------------------ what must be rewritten

    [Fact]
    public void AQualifiedWikiLinkFollowsTheFolder()
    {
        Assert.Equal("[[Assignments/Quiz 1]]", Rewrite("[[Tasks/Quiz 1]]"));
    }

    [Fact]
    public void ATransclusionFollowsTheFolder()
    {
        Assert.Equal("![[Assignments/diagram.png]]", Rewrite("![[Tasks/diagram.png]]"));
    }

    [Fact]
    public void AFolderDeepInAPathIsFound()
    {
        // A full vault path: ANY segment is a candidate, not only the first.
        Assert.Equal("[[ICS3U/section1/Assignments/Quiz 1]]",
                     Rewrite("[[ICS3U/section1/Tasks/Quiz 1]]"));
    }

    [Fact]
    public void AnAliasAndAHeadingAreLeftAlone()
    {
        // Both are the teacher's own words.
        Assert.Equal("[[Assignments/Quiz 1#Marking|the quiz]]",
                     Rewrite("[[Tasks/Quiz 1#Marking|the quiz]]"));
    }

    [Fact]
    public void AMarkdownLinkFollowsTheFolder()
    {
        Assert.Equal("[the quiz](Assignments/Quiz 1.md)", Rewrite("[the quiz](Tasks/Quiz 1.md)"));
        Assert.Equal("![](Assignments/diagram.png)", Rewrite("![](Tasks/diagram.png)"));
    }

    [Fact]
    public void APercentEncodedSegmentStaysEncoded()
    {
        // Obsidian writes Markdown links percent-encoded; handing it back a
        // decoded path would give it a link it cannot follow.
        Assert.Equal("[q](All%20Assignments/Quiz%201.md)",
                     FolderPathRewriter.Rewritten("[q](All%20Tasks/Quiz%201.md)", "All Tasks", "All Assignments"));
    }

    [Fact]
    public void ARelativePathIsStillRewritten()
    {
        Assert.Equal("[q](./Assignments/Quiz 1.md)", Rewrite("[q](./Tasks/Quiz 1.md)"));
    }

    [Fact]
    public void MatchingIgnoresCase()
    {
        // Windows filesystems are case-insensitive but case-preserving, so a
        // link written "tasks/" names the same folder.
        Assert.Equal("[[Assignments/Quiz 1]]", Rewrite("[[tasks/Quiz 1]]"));
    }

    // -------------------------------------------- what must NOT be rewritten

    [Fact]
    public void ABarePageLinkIsUntouched()
    {
        // The fact that makes this feature safe: Obsidian resolves a bare link
        // by searching the vault, so moving the folder leaves it working.
        Assert.Equal("[[Quiz 1]]", Rewrite("[[Quiz 1]]"));
    }

    [Fact]
    public void APageNamedAfterTheFolderIsUntouched()
    {
        // The match stops before the last segment, so a FILE called Tasks.md
        // is never a candidate.
        Assert.Equal("[[Tasks]]", Rewrite("[[Tasks]]"));
        Assert.Equal("[[Unit 1/Tasks]]", Rewrite("[[Unit 1/Tasks]]"));
        Assert.Equal("[q](Unit 1/Tasks.md)", Rewrite("[q](Unit 1/Tasks.md)"));
    }

    [Fact]
    public void AFolderWhoseNameMerelyContainsTheOldOneIsUntouched()
    {
        Assert.Equal("[[Extra Tasks/Quiz 1]]", Rewrite("[[Extra Tasks/Quiz 1]]"));
        Assert.Equal("[[Tasks Archive/Quiz 1]]", Rewrite("[[Tasks Archive/Quiz 1]]"));
    }

    [Fact]
    public void PlainProseMentioningTheFolderIsUntouched()
    {
        Assert.Equal("Put your work in the Tasks folder, in Tasks/ if you like.",
                     Rewrite("Put your work in the Tasks folder, in Tasks/ if you like."));
    }

    [Fact]
    public void AWebAddressIsNeverRewritten()
    {
        // The one that was got wrong on the mac: `Tasks` sits in this URL as an
        // ordinary segment, and the walk is blind to what a path MEANS.
        Assert.Equal("[handout](https://example.com/Tasks/handout.pdf)",
                     Rewrite("[handout](https://example.com/Tasks/handout.pdf)"));
    }

    [Fact]
    public void OtherSchemesAreLeftAloneToo()
    {
        Assert.Equal("[mail](mailto:someone@example.com/Tasks/x)",
                     Rewrite("[mail](mailto:someone@example.com/Tasks/x)"));
        Assert.Equal("[o](obsidian://open?vault=v&file=Tasks/Quiz)",
                     Rewrite("[o](obsidian://open?vault=v&file=Tasks/Quiz)"));
    }

    [Fact]
    public void AnAbsolutePathOnThisMachineIsNeverRewritten()
    {
        Assert.Equal(@"[x](C:/Users/teacher/Tasks/Quiz 1.md)", Rewrite(@"[x](C:/Users/teacher/Tasks/Quiz 1.md)"));
        Assert.Equal("[x](/Users/teacher/Tasks/Quiz 1.md)", Rewrite("[x](/Users/teacher/Tasks/Quiz 1.md)"));
    }

    [Fact]
    public void RenamingToTheSameNameChangesNothing()
    {
        Assert.Equal("[[Tasks/Quiz 1]]", FolderPathRewriter.Rewritten("[[Tasks/Quiz 1]]", "Tasks", "Tasks"));
    }

    [Fact]
    public void NothingAtAllIsNotACrash()
    {
        Assert.Equal("", FolderPathRewriter.Rewritten("", "Tasks", "Assignments"));
        Assert.Equal("x", FolderPathRewriter.Rewritten("x", "", "Assignments"));
        Assert.Equal("x", FolderPathRewriter.Rewritten("x", "Tasks", ""));
    }

    // ---------------------------------------------------------- the counting

    [Fact]
    public void CountingFindsOnlyQualifiedLinks()
    {
        string page = "[[Tasks/Quiz 1]] and [[Quiz 2]] and [q](Tasks/Quiz%203.md) and [[Extra Tasks/Q]]";
        Assert.Equal(2, FolderPathRewriter.Count(page, "Tasks"));
    }

    [Fact]
    public void CountingIsZeroWhenNothingPointsIn()
    {
        Assert.Equal(0, FolderPathRewriter.Count("[[Quiz 1]] and some prose about Tasks", "Tasks"));
        Assert.Equal(0, FolderPathRewriter.Count("", "Tasks"));
    }

    [Fact]
    public void CountingIgnoresTheWeb()
    {
        Assert.Equal(0, FolderPathRewriter.Count("[h](https://example.com/Tasks/handout.pdf)", "Tasks"));
    }
}
